external-crates/move/crates/move-vm-runtime/src/interpreter/eval.rs

```diff
@@ -0,0 +1,1158 @@
+        state,
+        resolver,
+        extensions,
+        runtime_limits_config,
+        gas_meter.remaining_gas(),
+    );
+    let native_function = function.get_native()?;
+
+    gas_meter.charge_native_function_before_execution(
+        ty_args.iter().map(|ty| TypeWithLoader {
+            ty,
+            loader: resolver.loader(),
+        }),
+        args.iter(),
+    )?;
+
+    let result = native_function(&mut native_context, ty_args.to_vec(), args)?;
+
+    // Note(Gas): The order by which gas is charged / error gets returned MUST NOT be modified
+    //            here or otherwise it becomes an incompatible change!!!
+    let return_values = match result.result {
+        Ok(vals) => {
+            gas_meter.charge_native_function(result.cost, Some(vals.iter()))?;
+            vals
+        }
+        Err(code) => {
+            gas_meter
+                .charge_native_function(result.cost, Option::<std::iter::Empty<&Value>>::None)?;
+            return Err(PartialVMError::new(StatusCode::ABORTED).with_sub_status(code));
+        }
+    };
+
+    // Paranoid check to protect us against incorrect native function implementations. A native function that
+    // returns a different number of values than its declared types will trigger this check
+    if return_values.len() != return_type_count {
+        return Err(
+            PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR).with_message(
+                "Arity mismatch: return value count does not match return type count".to_string(),
+            ),
+        );
+    }
+    Ok(return_values)
+}
+
+/// Perform a binary operation to two values at the top of the stack.
+fn binop<F, T>(state: &mut MachineState, f: F) -> PartialVMResult<()>
+where
+    Value: VMValueCast<T>,
+    F: FnOnce(T, T) -> PartialVMResult<Value>,
+{
+    let rhs = state.pop_operand_as::<T>()?;
+    let lhs = state.pop_operand_as::<T>()?;
+    let result = f(lhs, rhs)?;
+    state.push_operand(result)
+}
+
+/// Perform a binary operation for integer values.
+fn binop_int<F>(state: &mut MachineState, f: F) -> PartialVMResult<()>
+where
+    F: FnOnce(IntegerValue, IntegerValue) -> PartialVMResult<IntegerValue>,
+{
+    binop(state, |lhs, rhs| {
+        Ok(match f(lhs, rhs)? {
+            IntegerValue::U8(x) => Value::u8(x),
+            IntegerValue::U16(x) => Value::u16(x),
+            IntegerValue::U32(x) => Value::u32(x),
+            IntegerValue::U64(x) => Value::u64(x),
+            IntegerValue::U128(x) => Value::u128(x),
+            IntegerValue::U256(x) => Value::u256(x),
+        })
+    })
+}
+
+/// Perform a binary operation for boolean values.
+fn binop_bool<F, T>(state: &mut MachineState, f: F) -> PartialVMResult<()>
+where
+    Value: VMValueCast<T>,
+    F: FnOnce(T, T) -> PartialVMResult<bool>,
+{
+    binop(state, |lhs, rhs| Ok(Value::bool(f(lhs, rhs)?)))
+}
+
+#[derive(PartialEq, Eq)]
+enum StepStatus {
+    Running,
+    Done,
+}
```

Comment:
nit, I usually find it helpful to have type declarations at the top of the file, but up to you. 
I just went looking for it as soon as I found it :)